c10/core/CPUCachingAllocator.h

```diff
@@ -21,40 +22,84 @@ class CPUCachingAllocator {
    * No speculative allocation for any future allocations.
    */
   private:
-    std::mutex mutex_;
     // Invariants.
-    // 1. If an allocated memory pointer exists in allocation_map_ then
-    //    it must not exist in available_map_, and vice-e-versa.
-    // 2. All the pointers in available_map_ must be unique.
-    ska::flat_hash_map<void*, size_t> allocation_map_;
-    ska::flat_hash_map<size_t, std::deque<void*>> available_map_;
+    // 1. If memory is ever allocated via this allocator then
+    //    the pointer will exist in allocation_map_, unless the allocator
+    //    returned the memory to OS via free_cached.
+    //  1.1. Therefore even when the said memory is "freed" via this
+    //       allocator (and thus cached), it will continue to stay
+    //       in allocaiton_map_. Furthermore it will also exist in
+    //       available_map_. Thus an allocated memory pointer can be in both
+    //       allocation_map_ and available_map_ simultaneously.
+    // 2. Memory pointer maybe removed from allocation_map_, when it
+    //    is freed outside of the scope of this allocator, but was allocated
+    //    by this allocator.
+    // 3. Available map only contains that memory which was allocated
+    //    by this allocator and subsequently freed by this allocator.
+    // As a result of above invariants, allocated memory ptr cannot be in
+    // available_map_ unless it is in allocation_map_ as well.
+    ska::flat_hash_map<size_t, c10::SmallVector<void*, 16>> available_map_;
+    static ska::flat_hash_map<void*, size_t> allocation_map_;
```

Comment:
I think the mix of static and non-static state makes the code a bit hard to follow.  This is probably fine for now since we're talking about breaking this up into two different classes to allow users to override behavior.  But if that plan falls through for whatever reason, I think we should come back and refactor this.

Reply:
OK, so this one I thought about it a bit. It seemed to me that this class is the natural place for it, since it is coupled with what the class does and what it implies. Basically it, allocaiton_map_, wants to keep the global information about everything that happens through any instance of this class. My main issue was should it be outside where I might have to encapsulate in yet another class, would be to decouple two things that have strong association with each other. So if that continues to be case, I would still think this will be the place for it.
I agree about it being little hard to follow, but to me that is ok. What do you think?

Reply:
I think it should be two separate classes, but let's leave it as-is for now.  If/when we allow the user to override the caching logic, we can re-evaluate.